File: List8/CPPCode/ClientHandler.cpp
#include "ClientHandler.hpp"
#include <iostream>
#include <sstream>
#include <vector>

#ifdef _WIN32
    #include <winsock2.h>
    #define close_socket closesocket
#else
    #include <sys/socket.h>
    #include <unistd.h>
    #define close_socket close
#endif

ClientHandler::ClientHandler(int socket, BT<int, std::string>& it, BT<double, std::string>& dt, BT<std::string, std::string>& st, BT<Point2D, std::string>& pt)
    : clientSocket(socket), intTree(it), doubleTree(dt), stringTree(st), pointTree(pt) {}

void ClientHandler::run() {
    std::string inputLine;
    while (!(inputLine = readLine()).empty()) {
        std::string response = handleCommand(inputLine);
        sendLine(response);
    }
    close_socket(clientSocket);
    std::cout << "Klient rozlaczony.\n";
}

std::string ClientHandler::readLine() {
    std::string line = "";
    char ch;
    while (recv(clientSocket, &ch, 1, 0) > 0) {
        if (ch == '\n') break;
        if (ch != '\r') line += ch;
    }
    return line;
}

void ClientHandler::sendLine(const std::string& line) {
    std::string fullLine = line + "\n";
    send(clientSocket, fullLine.c_str(), fullLine.length(), 0);
}

void ClientHandler::replaceAll(std::string& str, const std::string& from, const std::string& to) {
    size_t start_pos = 0;
    while ((start_pos = str.find(from, start_pos)) != std::string::npos) {
        str.replace(start_pos, from.length(), to);
        start_pos += to.length();
    }
}

std::string ClientHandler::handleCommand(const std::string& input) {
    if (input.empty()) return "ERROR: Pusta komenda";

    std::stringstream ss(input);
    std::string command, type, keyStr, value;

    ss >> command;
    for (char &c : command) c = toupper(c);

    try {
        if (command == "ADD") {
            ss >> type >> keyStr;
            std::getline(ss >> std::ws, value); 
            if (type.empty() || keyStr.empty() || value.empty()) return "ERROR: Skladnia: ADD [typ] [klucz] [wartosc]";
            
            for (char &char_type : type) char_type = toupper(char_type);

            if (type == "INTEGER") {
                intTree.insert(std::stoi(keyStr), value);
            } else if (type == "DOUBLE") {
                doubleTree.insert(std::stod(keyStr), value);
            } else if (type == "POINT") {
                size_t comma = keyStr.find(',');
                if (comma == std::string::npos) throw std::invalid_argument("Format to x,y");
                int px = std::stoi(keyStr.substr(0, comma));
                int py = std::stoi(keyStr.substr(comma + 1));
                pointTree.insert(Point2D(px, py), value);
            } else {
                stringTree.insert(keyStr, value);
            }
            return "OK: Dodano pomyslnie na drzewie " + type;
        } 
        else if (command == "GET") {
            ss >> type >> keyStr;
            if (type.empty() || keyStr.empty()) return "ERROR: Skladnia: GET [typ] [klucz]";
            
            for (char &char_type : type) char_type = toupper(char_type);
            std::string resVal;
            bool found = false;

            if (type == "INTEGER") {
                found = intTree.search(std::stoi(keyStr), resVal);
            } else if (type == "DOUBLE") {
                found = doubleTree.search(std::stod(keyStr), resVal);
            } else if (type == "POINT") {
                size_t comma = keyStr.find(',');
                if (comma == std::string::npos) throw std::invalid_argument("Format to x,y");
                int px = std::stoi(keyStr.substr(0, comma));
                int py = std::stoi(keyStr.substr(comma + 1));
                found = pointTree.search(Point2D(px, py), resVal);
            } else {
                found = stringTree.search(keyStr, resVal);
            }
            return found ? "VALUE: " + resVal : "NOT_FOUND: Brak klucza na drzewie " + type;
        } 
        else if (command == "DELETE") {
            ss >> type >> keyStr;
            if (type.empty() || keyStr.empty()) return "ERROR: Skladnia: DELETE [typ] [klucz]";
            
            for (char &char_type : type) char_type = toupper(char_type);

            if (type == "INTEGER") {
                intTree.remove(std::stoi(keyStr));
            } else if (type == "DOUBLE") {
                doubleTree.remove(std::stod(keyStr));
            } else if (type == "POINT") {
                size_t comma = keyStr.find(',');
                if (comma == std::string::npos) throw std::invalid_argument("Format to x,y");
                int px = std::stoi(keyStr.substr(0, comma));
                int py = std::stoi(keyStr.substr(comma + 1));
                pointTree.remove(Point2D(px, py));
            } else {
                stringTree.remove(keyStr);
            }
            return "OK: Usunieto z drzewa " + type;
        } 
        else if (command == "SHOW") {
            ss >> type;
            if (type.empty()) return "ERROR: Skladnia: SHOW [typ]";
            
            for (char &char_type : type) char_type = toupper(char_type);
            std::string structure;

            if (type == "INTEGER") {
                structure = intTree.toReadableString();
            } else if (type == "DOUBLE") {
                structure = doubleTree.toReadableString();
            } else if (type == "POINT") {
                structure = pointTree.toReadableString();
            } else {
                structure = stringTree.toReadableString();
            }

            replaceAll(structure, "\n", "<BR>");
            return "TREE: <BR>" + structure;
        }

        return "ERROR: Nieznana komenda. Dostepne: ADD, GET, DELETE, SHOW";
    } 
    catch (const std::invalid_argument& e) {
        return "ERROR: Blad formatu klucza! Upewnij sie, ze pasuje do wybranego typu (np. x,y dla POINT).";
    } 
    catch (const std::exception& e) {
        return "ERROR: Blad operacji: " + std::string(e.what());
    }
}

File: List8/JavaCode/ClientHandler.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class ClientHandler implements Runnable {
    private final Socket socket;
    private final BT<Integer> intTree;
    private final BT<Double> doubleTree;
    private final BT<String> stringTree;
    private final BT<Point2D> pointTree; 

    public ClientHandler(Socket socket, BT<Integer> intTree, BT<Double> doubleTree, BT<String> stringTree, BT<Point2D> pointTree) {
        this.socket = socket;
        this.intTree = intTree;
        this.doubleTree = doubleTree;
        this.stringTree = stringTree;
        this.pointTree = pointTree;
    }

    @Override
    public void run() {
        try (
            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter writer = new PrintWriter(socket.getOutputStream(), true)
        ) {
            String inputLine;
            while ((inputLine = reader.readLine()) != null) {
                String response = processCommand(inputLine);
                writer.println(response);
            }
        } catch (IOException e) {
            System.out.println("Klient rozłączony: " + socket.getRemoteSocketAddress());
        } finally {
            try {
                if (socket != null && !socket.isClosed()) socket.close();
            } catch (IOException ignored) {}
        }
    }

    private String processCommand(String input) {
        if (input == null || input.trim().isEmpty()) return "ERROR: Pusta komenda";

        String[] arguments = input.trim().split("\\s+", 2);
        String command = arguments[0].toUpperCase();
        String payload = arguments.length > 1 ? arguments[1] : "";

        try {
            switch (command) {
                case "ADD": return handleAdd(payload);
                case "GET": return handleGet(payload);
                case "DELETE": return handleDelete(payload);
                case "SHOW": return handleShow(payload);
                default: return "ERROR: Nieznana komenda. Dostępne: ADD, GET, DELETE, SHOW";
            }
        } catch (NumberFormatException e) {
            return "ERROR: Błąd formatu klucza! Użyj formatu x,y dla POINT.";
        } catch (Exception e) {
            return "ERROR: Operacja nie powiodła się: " + e.getMessage();
        }
    }

    private String handleAdd(String payload) {
        String[] parts = payload.split("\\s+", 2);
        if (parts.length < 2) return "ERROR: Składnia: ADD [typ] [klucz]";
        
        String type = parts[0].toUpperCase();
        String keyStr = parts[1];

        switch (type) {
            case "INTEGER": intTree.insert(Integer.parseInt(keyStr)); break;
            case "DOUBLE": doubleTree.insert(Double.parseDouble(keyStr)); break;
            case "POINT": pointTree.insert(parsePoint(keyStr)); break;
            default: stringTree.insert(keyStr); break;
        }
        return "OK: Dodano do drzewa " + type;
    }

    private String handleGet(String payload) {
        String[] parts = payload.split("\\s+", 2);
        if (parts.length < 2) return "ERROR: Składnia: GET [typ] [klucz]";
        
        String type = parts[0].toUpperCase();
        String keyStr = parts[1];
        boolean found;

        switch (type) {
            case "INTEGER": found = intTree.search(Integer.parseInt(keyStr)); break;
            case "DOUBLE": found = doubleTree.search(Double.parseDouble(keyStr)); break;
            case "POINT": found = pointTree.search(parsePoint(keyStr)); break;
            default: found = stringTree.search(keyStr); break;
        }
        return found ? "FOUND: Klucz " + keyStr + " istnieje." : "NOT_FOUND: Brak klucza.";
    }

    private String handleDelete(String payload) {
        String[] parts = payload.split("\\s+", 2);
        if (parts.length < 2) return "ERROR: Składnia: DELETE [typ] [klucz]";
        
        String type = parts[0].toUpperCase();
        String keyStr = parts[1];

        switch (type) {
            case "INTEGER": intTree.delete(Integer.parseInt(keyStr)); break;
            case "DOUBLE": doubleTree.delete(Double.parseDouble(keyStr)); break;
            case "POINT": pointTree.delete(parsePoint(keyStr)); break;
            default: stringTree.delete(keyStr); break;
        }
        return "OK: Usunięto z drzewa " + type;
    }

    private String handleShow(String payload) {
        String type = payload.trim().toUpperCase();
        if (type.isEmpty()) return "ERROR: Składnia: SHOW [typ]";

        String structure;
        switch (type) {
            case "INTEGER": structure = intTree.serialize(); break;
            case "DOUBLE": structure = doubleTree.serialize(); break;
            case "POINT": structure = pointTree.serialize(); break;
            default: structure = stringTree.serialize(); break;
        }
        return "TREE: " + structure;
    }

    private Point2D parsePoint(String keyStr) {
        String[] coords = keyStr.split(",");
        if (coords.length != 2) throw new NumberFormatException();
        return new Point2D(Integer.parseInt(coords[0]), Integer.parseInt(coords[1]));
    }
}